rust-manifest/src/manifest.rs

```diff
@@ -48,21 +48,47 @@ pub struct Component {
     pub target: String,
 }
 
+// These secondary types are returned in operations on the manifest
+
+#[derive(Clone, Debug)]
+pub struct Diff {
+    pub packages: Vec<RequiredPackage>,
+    pub to_install: Vec<Component>,
+    pub to_uninstall: Vec<Component>,
+}
+
+#[derive(Clone, Debug)]
+pub struct RequiredPackage {
+    pub url: String,
+    pub hash: String,
+}
+
 impl Manifest {
-    fn get_packages(table: toml::Table, path: &str) -> Result<HashMap<String, Package>> {
-        let mut result = HashMap::new();
+    pub fn init(root_package: &str, target: &str) -> Self {
+        let mut result = Manifest {
+            manifest_version: DEFAULT_MANIFEST_VERSION.to_owned(),
+            date: String::new(),
+            root: Some(root_package.to_owned()),
+            packages: HashMap::new(),
+        };
 
-        for (k, v) in table {
-            if let toml::Value::Table(t) = v {
-                let path = format!("{}{}.", path, &k);
-                result.insert(k, try!(Package::from_toml(t, &path)));
-            }
-        }
+        result.packages.insert(root_package.to_owned(), Package::init(target));
 
-        Ok(result)
+        result
+    }
+
+    pub fn parse(data: &str) -> Result<Self> {
+        let mut parser = toml::Parser::new(data);
+        let value = try!(parser.parse().ok_or_else(move || Error::Parsing(parser.errors)));
+
+        Self::from_toml(value, "")
```

Comment:
In terms of ease of converting TOML -> Rust structures you may also want to look into using the `RustcEncodable` and `RustcDecodable` support. The support in TOML should provide nice error messages as well in terms of "expected a table but found a string". FWIW the `Decodable` support is what Cargo uses to deserialize a manifest into a Rust structure.

I've also found the Rust "schema" as an easy way to understand how the document is structured as well.